### PR TITLE
DEV: routable chat part 2

### PR DESCRIPTION
**Pull request by jjaffeux**
This commit is expanding on previous work making everything chat working through an URL.

- Improves drawer templates to be all URLs
- Implements some kind of router for the drawer
- Removes few remaining actions for opening channels

### CODE REVIEW

**Comment by jjaffeux** [comment]
@martin-brennan I will finish this tomorrow, but pushing now, so you are aware of it. That should make much easier to plug thread into drawer when we want to.

File: plugins/chat/assets/javascripts/discourse/components/chat-live-pane.js [hunk 1]
```diff
@@ -1056,7 +1055,7 @@ export default Component.extend({
           upload_ids: (uploads || []).mapBy("id"),
         },
       }).then(() => {
-        this.onSwitchChannel(c);
+        this.router.transitionTo("chat.channel", "-", c.id);
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-live-pane.js:1058]: This is one of the big advantages of the work of the last months, everything is routable, we don't need all these actions everywhere, an URL is all you need!

File: plugins/chat/assets/javascripts/discourse/components/chat-drawer.hbs [hunk 2]
```diff
@@ -10,128 +11,14 @@
...
-            />
-          {{else}}
-            <ChannelsList
-              @onOpenView={{action "openURL"}}
-              @onSelect={{action "switchChannel"}}
-            />
-          {{/if}}
-        </div>
-      {{/if}}
+      <this.chatDrawerRouter.component
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/components/chat-drawer.hbs:14]: Had no idea you could do this, nice

File: plugins/chat/assets/javascripts/discourse/components/chat-drawer.js [hunk 3]
```diff
@@ -109,35 +84,19 @@ export default Component.extend({
...
-  },
-
   computeDrawerStyle() {
     const { width, height } = this.chatDrawerSize.size;
     let style = `width: ${escapeExpression((width || "0").toString())}px;`;
     style += `height: ${escapeExpression((height || "0").toString())}px;`;
     this.set("drawerStyle", htmlSafe(style));
   },
 
-  openChannelAtMessage(channel, messageId) {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/components/chat-drawer.js:139]: Really nice that you got rid of the openChannel stuff :+1:

File: plugins/chat/assets/javascripts/discourse/components/chat-drawer.js [hunk 4]
```diff
@@ -204,74 +163,22 @@ export default Component.extend({
...
   },
 
-  _highlightCb(messageId) {
-    if (messageId) {
-      return () => {
-        this.appEvents.trigger("chat-live-pane:highlight-message", messageId);
-      };
-    }
-  },
+  _routeFromURL(URL) {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/components/chat-drawer.js:173]: nit, but I think it's odd to use `URL` (all caps) for something that's not a constant, should just be `url`

File: plugins/chat/assets/javascripts/discourse/components/chat-live-pane.js [hunk 5]
```diff
@@ -986,57 +985,58 @@ export default Component.extend({
...
-    let data = {
-      message,
-      cooked,
-      staged_id: stagedId,
-      upload_ids: uploads.map((upload) => upload.id),
-    };
-    if (this.replyToMsg) {
-      data.in_reply_to_id = this.replyToMsg.id;
-    }
+    return this.chat.loadCookFunction(this.site.categories).then((cook) => {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/components/chat-live-pane.js:988]: I hope we can move this out of live pane at some point too, surely we can just do this once. Noticed this when I was digging in yesterday

**Review by martin-brennan — APPROVED** [review]
Like the changes made here :heart: Just a couple of comments mostly clarifications.

File: plugins/chat/assets/javascripts/discourse/components/chat-live-pane.js (hunk 5 above)
Comment (jjaffeux) [reply plugins/chat/assets/javascripts/discourse/components/chat-live-pane.js:988]: Yes totally agreed

File: plugins/chat/assets/javascripts/discourse/services/chat-drawer-router.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    extractParams: (route) => {
+      return {
+        channelId: route.params.channelId,
+        messageId: route.queryParams.messageId,
+      };
+    },
+  },
+};
+
+export default class ChatDrawerRouter extends Service {
```
[the remark this answers is not in the document]
Comment (jjaffeux) [reply plugins/chat/assets/javascripts/discourse/services/chat-drawer-router.js:31]: The thing is ChatDrawer is a component and not tied to a real route. So here I am just faking it so we have some kind of 1-1 logic with the full page chat routing logic. I think it makes things easier to reason about.

File: plugins/chat/assets/javascripts/discourse/components/chat-drawer.js (hunk 4 above)
Comment (jjaffeux) [reply plugins/chat/assets/javascripts/discourse/components/chat-drawer.js:173]: yes, it's some habit I have, no big deals, I can change it :D

**Pull request merged**